src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,268 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+                //    |     \--*  CNS_INT   int    -8 (offset)
+                //    \--*  CNS_INT   int    0
+                //
+                GenTree* isInitAdrNode;
+                if (isInitOffset != 0)
+                {
+                    // Don't fold ADD(CNS1, CNS2) here since the result won't be reloc-friendly for AOT
+                    isInitAdrNode =
+                        gtNewIndir(TYP_INT, gtNewOperNode(GT_ADD, TYP_I_IMPL,
+                                                          gtNewIconHandleNode(staticInitAddr, GTF_ICON_CONST_PTR),
+                                                          gtNewIconNode(isInitOffset)));
+                    isInitAdrNode->gtFlags &= ~GTF_EXCEPT;
+                    isInitAdrNode->gtFlags |= (GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+                }
+                else
+                {
+                    isInitAdrNode = gtNewIndOfIconHandleNode(TYP_INT, staticInitAddr, GTF_ICON_CONST_PTR, true);
+                }
+
+                // Don't emit mask if it's 0xFFFFFFFF, although, JIT should be able to drop it as redundant itself
+                if (isInitMask < 0xFFFFFFFF)
+                {
+                    isInitAdrNode = gtNewOperNode(GT_AND, TYP_INT, isInitAdrNode, gtNewIconNode(isInitMask));
+                }
+
+                GenTree* isInitedCmp = gtNewOperNode(GT_NE, TYP_INT, isInitAdrNode, gtNewIconNode(0));
+                isInitedCmp->gtFlags |= GTF_RELOP_JMP_USED;
+                BasicBlock* isInitedBb =
+                    fgNewBBFromTreeAfter(BBJ_COND, prevBb, gtNewOperNode(GT_JTRUE, TYP_VOID, isInitedCmp), debugInfo);
+
+                // Fallback basic block
+                // TODO-CQ: for JIT we can replace the original call with CORINFO_HELP_INITCLASS
+                // that only accepts a single argument
+                BasicBlock* helperCallBb = fgNewBBFromTreeAfter(BBJ_NONE, isInitedBb, call, debugInfo, true);
+
+                //
+                // Update preds in all new blocks
+                //
+
+                // Unlink block and prevBb
+                fgRemoveRefPred(block, prevBb);
+
+                // Block has two preds now: either isInitedBb or helperCallBb
+                fgAddRefPred(block, isInitedBb);
+                fgAddRefPred(block, helperCallBb);
+
+                // prevBb always flow into isInitedBb
+                fgAddRefPred(isInitedBb, prevBb);
+
+                // Both fastPathBb and helperCallBb have a single common pred - isInitedBb
+                fgAddRefPred(helperCallBb, isInitedBb);
+
+                // helperCallBb unconditionally jumps to the last block (jumps over fastPathBb)
+                isInitedBb->bbJumpDest = block;
+
+                //
+                // Re-distribute weights
+                //
+
+                block->inheritWeight(prevBb);
+                isInitedBb->inheritWeight(prevBb);
+                helperCallBb->bbSetRunRarely();
```

Comment:
Thoughts about introducing method with a different name for this purpose?

https://github.com/dotnet/runtime/pull/83911/#discussion_r1148172248

Reply:
It's possible it will never be executed, right? if the type was initialized in a different method. I didn't change it/introduce a new API because I know that @AndyAyersMS is considering to introduce that concept now (zero weight vs "almost zero, but not realy" weight)

Reply:
Sounds good.

> It's possible it will never be executed, right?

Yes. It is also possible that it will be executed more than once (e.g. when there are multiple threads racing to run the same cctor). It is why I have suggested `bbSetRunTypicallyOnce` as the name.